Sources/core/objects/Revolver.cpp

```diff
@@ -0,0 +1,67 @@
+#include "Revolver.h"
+
+Revolver::Revolver(size_t id) : Item(id)
+{
+    SetSprite("icons/guns.dmi");
+    SetState("revolver");
+
+    name = "Revolver",
+    
+    ammunition_ = 6;
+}
+
+bool Revolver::Handle_Ammo()
+{
+    if (ammunition_ <= 0)
+    {
+        ammunition_ = 0;
+        return false;
+    }
+    --ammunition_;
+    return true;
+}
+
+void Revolver::Shoot(sf::Vector2i target)
+{	  
+    if(Handle_Ammo())
+    {
+        id_ptr_on<Projectile> p = Create<Projectile>(Bullet::T_ITEM_S(),GetOwner());
```

Comment:
`id_ptr_on` just a holder for some object (smart pointer). Most of objects have owners, exceptions are rare. If object was clicked then it should have owner. For the most object owner is a some tile (note that tile and turf are different concepts in Griefly).
About `AttackBy` - I was wrong, my mistake. Indeed, the `AttackBy` argument is `id_ptr_on<Item>`, so there are some additional actions needed.
Probably it is better to use `Bump` function, because it has `id_ptr_on<IMovable>` argument. So humans will be kind of "bumped" by projectiles.